geopandas/tools/clip.py

```diff
@@ -37,41 +50,94 @@ def _clip_points(gdf, poly):
         The returned GeoDataFrame is a subset of gdf that intersects
         with poly.
     """
-    return gdf.iloc[gdf.sindex.query(poly, predicate="intersects")]
+    clipping_candidates = gdf.iloc[gdf.sindex.query(poly, predicate="intersects")]
+    point_indices = clipping_candidates.geom_type == "Point"
+    # For performance reasons points don't need to be intersected with poly
+    points = clipping_candidates[point_indices]
 
+    non_points = clipping_candidates[~point_indices]
+    non_points_intersected = _intersect_non_point_geometries_with_poly(non_points, poly)
+    return pd.concat([points, non_points_intersected])
 
-def _clip_line_poly(gdf, poly):
-    """Clip line and polygon geometry to the polygon extent.
 
-    Clip an input line or polygon to the polygon extent of the poly
-    parameter. Parts of Lines or Polygons that intersect the poly geometry are
-    extracted with associated attributes and returned.
+def _validate_inputs(gdf, mask):
+    if not isinstance(gdf, (GeoDataFrame, GeoSeries)):
+        raise TypeError(
+            "'gdf' should be GeoDataFrame or GeoSeries, got {}".format(type(gdf))
+        )
 
-    Parameters
-    ----------
-    gdf : GeoDataFrame, GeoSeries
-        Line or polygon geometry that is clipped to poly.
+    if not isinstance(mask, (GeoDataFrame, GeoSeries, Polygon, MultiPolygon)):
+        raise TypeError(
+            "'mask' should be GeoDataFrame, GeoSeries or"
+            "(Multi)Polygon, got {}".format(type(mask))
+        )
+    if isinstance(mask, (GeoDataFrame, GeoSeries)):
+        if not _check_crs(gdf, mask):
+            _crs_mismatch_warn(gdf, mask, stacklevel=3)
 
-    poly : (Multi)Polygon
-        Reference polygon for clipping.
 
-    Returns
-    -------
-    GeoDataFrame
-        The returned GeoDataFrame is a clipped subset of gdf
-        that intersects with poly.
-    """
-    gdf_sub = gdf.iloc[gdf.sindex.query(poly, predicate="intersects")]
+def _keeping_geometry_type_is_possible(gdf):
+    df_contains_geometry_collection = (gdf.geom_type == "GeometryCollection").any()
+    if df_contains_geometry_collection:
+        warnings.warn(
+            "keep_geom_type can not be called on a GeoDataFrame with "
+            "GeometryCollection."
+        )
+        return False
+    number_of_unique_geometry_types = sum(
+        gdf.geom_type.isin(geom_type).any() for geom_type in _GEOMETRY_TYPES
+    )
+    if number_of_unique_geometry_types > 1:
```

Comment:
```suggestion
    if sum(gdf.geom_type.isin(geom_type).any() for geom_type in _GEOMETRY_TYPES) > 1:
```

Reply:
I must admit that I disagree with removing these variables. I put them in intentionally as I believe they make it easier for a future reader. 

I tried to follow known best practices (such as the ones presented in Clean Code by Robert C. Martin) to the best of my skill and knowledge. The main reason for these variables is that, in my humble opinion, they improve the readability a lot. As a reader of this code, I care mostly about business logic on this level of abstraction and not so much about the low level implementation. 

E.g. I would argue that `sum(gdf.geom_type.isin(geom_type).any() for geom_type in _GEOMETRY_TYPES)` is a fairly complicated statement (where both of us had troubles understanding the difference to `.nunique()`) but if I read `number_of_unique_geometry_types` I know what's going on and I probably don't need to check how it is computed.

That being said, it's your code, so feel free to change it any way you want.

Reply:
Also, nevermind - I'll just change it in the next couple of days.

Reply:
What about a comment? We won't do it as a function but write an explanatory one-line comment above the code. That way we may both be happy? There'll be a clear explanation of what is going on without the need to fetch the definition of a function from elsewhere. 

I don't want to force you into anything you're not comfortable with, this is a community project and we should find an agreement. There's no space for enforcing anything.

Reply:
Oh, I think I am missing something here - which function are you talking about? 

Also not feeling forced :-) - it was quite the big refactoring and I am thankful for the thorough review. Considering the size of all the other things that change here, I don't want this to be held back by a discussion about such a minor, and very subjective, detail.

I will make the requested changes tomorrow 👍

Reply:
No function 🤦 , I meant variables, my brain didn't apparently work properly yesterday. 

Looking at these two suggestions again, I would go with the one above (`(gdf.geom_type == "GeometryCollection").any()`) because it is quite clear what this line does, there's no need to further _explain_ it by a variable name.

Feel free to keep the second one. Or add a comment. Up to you.

Reply:
Agreed! I made the necessary changes :-)